### PR TITLE
Anonymization/v1.0

### PR DESCRIPTION
**Pull request by AJun01**
# Anonymization Feature v1.0

## Introduction

Releasing of the **Anonymization Feature v1.0**. This feature is designed to enhance the privacy and security of our users by filtering out sensitive information in generated HTTP requests. 

## Key Features

- **Email Anonymization**
- **Name Anonymization**
- **Address Anonymization**

## How It Works

The Anonymization Feature v1.0 utilizes advanced regular expressions to detect sensitive information in HTTP requests. When enabled using setAnonymization(), it processes the request body and replaces the detected sensitive information with anonymized placeholders.

### Example

**Original Request Body**:
```
{
  "email": "YujunLiu@example.com",
  "name": "Yujun Liu",
  "address": "22 Bond str"
}
```

**Anonymized Request Body**:
```
{
  "email": "************@email.com",
  "name": "*Anonymized* Name",
  "address": "*Anonymized* Address"
}
```

## How to Enable

To enable the Anonymization Feature, simply call the `setAnonymization(true)` function in your configuration. This will ensure that all HTTP requests are processed to filter out sensitive information before being sent.

### Code Example

```javascript
const { setAnonymization } = require('../../lib/anonymizationProxy/anonymizer');

function run() {
    setAnonymization(true); 
}
```

## Test

image

##How to create a test: 

- create a mock server:
```const http = require('http');

const server = http.createServer((req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end(`Hello Yujun, your email is yujunliu@example.com, and your address is 22 bond str`);
});

server.listen(4000, () => {
    console.log('Mock server running on port 4000');
});
```

- create a test JS file and modify the function into:
```var requestFunc = function() {
    var options = {};
    options['host'] = 'localhost';
    options['port'] = '4000';
    options['path'] = '/';
    options['method'] = 'GET';

    var req = trafficSimulator.request(options, function(response) {
        console.log("Response: %s", response.statusCode);
        response.setEncoding('utf8');
        let responseBody = '';
        response.on('data', function(chunk) {
            responseBody += chunk;
        });
        response.on('end', function() {
            responseBody = anonymizeHttpResponse(response, responseBody);
            console.log('Anonymized Response:', responseBody);
        });
    });
}
``` 
## Yujun 6/16/2024

### CODE REVIEW

**Comment by iskitsas** [comment]
@AJun01 it seems like you've checked out code from the branch used in PR #66. Have you updated your branch to be up-to-date with the branch from PR #66?

File: compose-dev.yaml
Comment (iskitsas) [inline compose-dev.yaml:1]: Why did you introduce this file?

**Comment by AJun01** [comment]
yes the newest commits pushed to PR 66 were pushed to this PR as well

File: lib/AnonymizationProxy/TAS.js
Comment (iskitsas) [inline lib/AnonymizationProxy/TAS.js:1]: @AJun01 as agreed in our discussions and pointed it out in issue #73 , we should implement this as a module in the core library. Here you are creating a proxy server. This is not going to work and it is out of scope.

File: example/anonumized-request/anonymized-request.js
Comment (iskitsas) [inline example/anonumized-request/anonymized-request.js:1]: Folder name has a typo

File: lib/AnonymizationProxy/anonymizer.js
Comment (iskitsas) [inline lib/AnonymizationProxy/anonymizer.js:1]: Provide some unit tests to validate that anonymizer works as expected. Sample body requests:
json:
```json
{
  "email": "YujunLiu@example.com",
  "name": "Yujun Liu",
  "address": {
    "street": "22 Bond St",
    "city": "New York",
    "state": "NY",
    "zip": "10012",
    "country": "USA"
  },
  "phone": "+1-555-123-4567",
  "date_of_birth": "1990-05-15",
  "gender": "Male",
  "occupation": "Software Engineer",
  "company": {
    "name": "Tech Solutions Inc.",
    "address": {
      "street": "1500 Broadway",
      "city": "New York",
      "state": "NY",
      "zip": "10036",
      "country": "USA"
    },
    "position": "Senior Developer"
  },
  "social_media": {
    "linkedin": "https://linkedin.com/in/yujunliu",
    "twitter": "https://twitter.com/yujunliu"
  },
  "hobbies": ["coding", "reading", "hiking"],
  "education": {
    "degree": "Bachelor of Science in Computer Science",
    "university": "New York University",
    "graduation_year": 2012
  },
  "skills": [
    "JavaScript",
    "Python",
    "Java",
    "SQL",
    "React",
    "Node.js"
  ],
  "certifications": [
    {
      "name": "Certified Java Developer",
      "issuing_organization": "Oracle",
      "issue_date": "2015-06-01"
    },
    {
      "name": "AWS Certified Solutions Architect",
      "issuing_organization": "Amazon",
      "issue_date": "2017-11-01"
    }
  ]
}
```

form data:
```
--boundary123
Content-Disposition: form-data; name="first_name"

John
--boundary123
Content-Disposition: form-data; name="last_name"

Doe
--boundary123
Content-Disposition: form-data; name="email"

john.doe@example.com
--boundary123
Content-Disposition: form-data; name="password"

securepassword123
--boundary123
Content-Disposition: form-data; name="profile_picture"; filename="profile.jpg"
Content-Type: image/jpeg

<binary data of profile.jpg>
--boundary123--
```

File: lib/AnonymizationProxy/anonymizer.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+const emailRegex = /([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6})/g;
+const nameRegex = /\b([A-Z][a-z]+\s[A-Z][a-z]+)\b/g;
```
Comment (iskitsas) [inline lib/AnonymizationProxy/anonymizer.js:2]: If the body has any words matching this pattern, will you anonymize? It does not seem right

**Review by iskitsas — CHANGES_REQUESTED** [review]
Please revisit this PR and ensure the following:

1. Integrate the anonymization feature as a module in the core library. Find the appropriate place in the existing code to use your anonymization function if the feature is enabled.
2. Provide unit tests with various inputs to verify that your implementation works as expected without compromising any interchange format (e.g., JSON, multipart, etc.).

File: compose-dev.yaml
Comment (AJun01) [reply compose-dev.yaml:1]: automatic change made by Docker, I am removing it

File: lib/AnonymizationProxy/TAS.js
Comment (AJun01) [reply lib/AnonymizationProxy/TAS.js:1]: The proxy server was created as one of my implementation test idea, not involved in the  anonymization function v1.0, I will remove it ASAP

**Review by AJun01 — COMMENTED** [review]
Thank you for reviewing, fixing now

**Comment by AJun01** [comment]
Closing this PR to open a new one with the renamed branch and PR.

**Pull request closed without merging**